Use ViewModel & LiveData for Plugin Directory

oguzkocer opened the pull request:
This PR started as an experiment during a meetup. I've wanted to play with the `ViewModel` and `LiveData` components for a while now and even though, I did so in test apps, it was hard to tell whether it's good enough to use it or not without trying it on a complicated feature. Spoiler alert, I think it's not 100% ready, but still a lot better than our previous implementation, so I'd strongly prefer us to start using it.

I didn't know if I was going to open this PR and I didn't mean to include any bug fixes in it, however, I ended up fixing a bunch of small (edge case type) issues especially related to the search feature on plugin directory just because it was so easy to do so 😄. I've not kept a list of these issues unfortunately, so I can't really list them here easily. Having said that, I've tested the PR extensively and also asked around a bit about the implementation, so I am mostly comfortable with the changes.

@theck13 As we discussed during the meetup, I was hoping you could be the official reviewer on this PR since @nbradbury doesn't have time for it. However, let's wait on merging it (or even reviewing, it's up to you) before we get a 👍 from the team. I'll write a p2 post about this soon where we can decide whether to start using these components or not. I'll include my analysis of the components in the p2 post, so that should also help with the review!

oguzkocer commented:
Something I forgot to mention in the original PR description was that I opted not to use the HashMaps for caching the `WPOrgPlugin`s for the site plugins. I've played with this a lot and found that it was rarely useful and my opinion is that we should do these type of optimizations in FluxC side. That way it'd be much more reliable because FluxC knows more than the activity/fragment/viewmodels and we won't need to duplicate any effort. Also, if that turns out to be a pattern we want to utilize, we can use it for other features as well.

So, my suggestion is that, we test the feature without it first. If we feel we really need a cache, we open an issue in FluxC and start working on it immediately. If we think it'd be good but not crucial, we open an issue and leave it. If we are OK with this as is and don't really feel the performance issues, we ignore it, because I suspect we'd do such optimizations for all features together whenever we decide to do so.

@nbradbury This was something you worked on (albeit as a result of a comment), let me know what you think about it!

oguzkocer commented:
A quick note about the 27b5a64af76aa269b29d7caa25ef9ef05c6a3d65, after merging @nbradbury's changes, I realized that if I am in "Popular" plugins section and I rotate the device, the title is changed to the default. In the base branch, there are a lot of setTitle calls, so this wasn't a problem before. After trying out a few different approaches, I really like this one as it's very clean and all the setTitle calls are when we actually need it.

nbradbury commented:
> So, my suggestion is that, we test the feature without it first. If we feel we really need a cache, we open an issue in FluxC and start working on it immediately.

Honestly, I don't think it's a big deal - I simply mentioned it because the HashMaps were added at the suggestion of an earlier reviewer.

nbradbury commented:
I'm seeing a reproducible crash in this branch that doesn't occur in the master branch.

* Search for "spam" (or anything else)
* Tap a plugin to show the detail
* Boom!

```
FATAL EXCEPTION: main
Process: org.wordpress.android.beta, PID: 9075
java.lang.IllegalStateException: Two different ViewHolders have the same stable ID. Stable IDs in your adapter MUST BE unique and SHOULD NOT change.
ViewHolder 1:ViewHolder{2a75907 position=0 id=0, oldPos=-1, pLpos:-1}
View Holder 2:ViewHolder{ce03af4 position=6 id=0, oldPos=-1, pLpos:-1} android.support.v7.widget.RecyclerView{311f71d VFED..... ......ID 0,0-1080,1584 #7f0903f9 app:id/recycler}, adapter:org.wordpress.android.ui.plugins.PluginListFragment$PluginListAdapter@55f3d92, layout:android.support.v7.widget.LinearLayoutManager@f3be363, context:org.wordpress.android.ui.plugins.PluginBrowserActivity@e70dc73
at android.support.v7.widget.RecyclerView.handleMissingPreInfoForChangeError(RecyclerView.java:3826)
at android.support.v7.widget.RecyclerView.dispatchLayoutStep3(RecyclerView.java:3751)
at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3421)
at android.support.v7.widget.RecyclerView.onLayout(RecyclerView.java:3962)
```

My guess is a dup plugin is being added to the PluginListFragment's adapter.

theck13 commented on WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginBrowserActivity.java:
```diff
@@ -161,27 +114,69 @@ public void onClick(View v) {
+        mViewModel.getSitePluginsListStatus().observe(this, new Observer<PluginBrowserViewModel.PluginListStatus>() {
+            @Override
+            public void onChanged(@Nullable PluginBrowserViewModel.PluginListStatus listStatus) {
+                showProgress(listStatus == PluginBrowserViewModel.PluginListStatus.FETCHING
```
This also creates a synthetic accessor method.  We can make the `showProgress` method `protected` or package-private to avoid that.

oguzkocer replied:
All synthetic accessor method issues should be fixed in 8244bd59d148e7e560c82951daded148e249208f.

theck13 commented on WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginBrowserActivity.java:
```diff
@@ -161,27 +114,69 @@ public void onClick(View v) {
+            @Override
+            public void onChanged(@Nullable PluginBrowserViewModel.PluginListStatus listStatus) {
+                showProgress(listStatus == PluginBrowserViewModel.PluginListStatus.FETCHING
+                        && mViewModel.isSitePluginsEmpty());
```
The `mViewModel` reference also creates a synthetic accessor method.  We can make it `protected` or package-private to avoid that.

oguzkocer replied:
All synthetic accessor method issues should be fixed in 8244bd59d148e7e560c82951daded148e249208f.

theck13 commented on WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginListFragment.java:
```diff
@@ -213,14 +215,23 @@ void onLoadedMore(boolean canLoadMore) {
             }
         }
 
+        private void reloadPluginWithSlug(@NonNull String slug) {
```
This is another synthetic accessor method creation.  We can make the `reloadPluginWithSlug` method protected or package-private to avoid that.

theck13 commented on WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginListFragment.java:
```diff
@@ -74,30 +59,100 @@ public static PluginListFragment newInstance(@NonNull SiteModel site, @NonNull P
+        mViewModel.getSitePlugins().observe(this, new Observer<List<SitePluginModel>>() {
+            @Override
+            public void onChanged(@Nullable final List<SitePluginModel> sitePlugins) {
+                if (mListType == PluginListType.SITE) {
```
This is another private member reference outside of the private scope creating a synthetic accessor method adding to our app method count.  We can make `mListType` protected or package-private to avoid that.

oguzkocer replied:
All synthetic accessor method issues should be fixed in 8244bd59d148e7e560c82951daded148e249208f.

theck13 commented on WordPress/src/main/java/org/wordpress/android/viewmodel/PluginBrowserViewModel.java:
```diff
@@ -0,0 +1,411 @@
+                @Override
+                public void run() {
+                    if (StringUtils.equals(query, getSearchQuery())) {
+                        submitSearch(query, false);
```
We can make `submitSearch` protected or package-private to avoid creating synthetic accessor methods.

oguzkocer replied:
All synthetic accessor method issues should be fixed in 8244bd59d148e7e560c82951daded148e249208f.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/viewmodel/PluginBrowserViewModel.java:
```diff
@@ -0,0 +1,395 @@
+
+    private final Handler mHandler;
+
+    private MutableLiveData<PluginListStatus> mNewPluginsListStatus;
```
Fixed in 913bd2531ccb4dd8a7f8adbc80abf77e98c4387d. I've also added `@SuppressWarnings("WeakerAccess")` to the constructor because we need it to be `public` for the view model to work.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/viewmodel/PluginBrowserViewModel.java:
```diff
@@ -0,0 +1,395 @@
+    }
+
+    public SitePluginModel getSitePluginFromSlug(String slug) {
+        return mPluginStore.getSitePluginBySlug(getSite(), slug);
```
I did not see this comment when I wrote [this](https://github.com/wordpress-mobile/WordPress-Android/pull/7229#issuecomment-365719942), it was just a coincidence. I added your suggestion in 0aac57a2df5bc18679f9ac9fe50f750851f14af7 with a `TODO` to remove it if we add caching to `PluginStore`. As you said, if we still have the information, there is no reason not to use it here. It doesn't introduce another `HashMap` for us to juggle, so a nice middle ground. Having said that, if I can find time for it, I do want to look into how we can cache the information in FluxC because I think this feature can greatly benefit from it.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginBrowserActivity.java:
```diff
@@ -227,59 +235,15 @@ public boolean onOptionsItemSelected(final MenuItem item) {
     }
 
-    private void reloadPlugins(@NonNull PluginListType pluginType) {
+    private void refreshPluginAdapterAndVisibility(@NonNull PluginListType pluginType, List<?> plugins) {
```
I've gone back and forth on this one. My comment was about refreshing the actual data, when I saw `refreshPlugins` for example, I immediately thought fetch, which I guess I shouldn't have. The reason I went with `refresh` in this method was it felt weird to say `reloadVisibility` and it wasn't refreshing/reloading the actual data, but the adapter. Suffice to say, I was not happy either way, but couldn't find anything better.

Made the change in c7db7c8140c29e1e45b7c404940c57e0389fa548.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginBrowserActivity.java:
```diff
@@ -125,13 +81,10 @@ public void onCreate(Bundle savedInstanceState) {
 
-        if (mSite == null) {
+        if (mViewModel.getSite() == null) {
             ToastUtils.showToast(this, R.string.blog_not_found, ToastUtils.Duration.SHORT);
```
Fixed in 26319f5f28245dc0f059536c4dfdef9a28cada2c.

oguzkocer commented:
@nbradbury @theck13 The PR is ready for a second pass. Here is a list of all the changes made after the reviews to make that second pass easier:

* Addressed @nbradbury's feedback in 913bd2531ccb4dd8a7f8adbc80abf77e98c4387d, 0aac57a2df5bc18679f9ac9fe50f750851f14af7, 26319f5f28245dc0f059536c4dfdef9a28cada2c, 913bd2531ccb4dd8a7f8adbc80abf77e98c4387d. Each comment has the relevant commit, so you can check them as well.
* Addressed @theck13's feedback about the synthetic accessor methods in 8244bd59d148e7e560c82951daded148e249208f.
* While making these changes I realized that I forgot a todo for moving the fetch call to the `ViewModel` which is addressed in b881499a0a3d738834f98517a5f3ff872930593f.
* Since we removed the search icon from the fragments, we don't really need the `setListType` or `getListFragment` calls anymore. To keep things cleaner, I removed those in 5bd88715fba562d15d4693ea3c14c842c16741c5.
* Although extremely rare, I found a bug while testing the search. If I am not mistaken if the underlying data for the adapter changes exactly when the user taps on a plugin to open details page, the `getItem` method can return a null object. I added a null check for these in eec4d3e9238bb1fff06141de47291d681cceb711 which has the added benefit of getting rid of some of the AS warnings for it.
* Fixed the crash @nbradbury mentioned [here](https://github.com/wordpress-mobile/WordPress-Android/pull/7229#issuecomment-365752686) in ebde927e8337ca8ce79a7df1dcbf219e898cef71. The issue there was, the search results don't have an id because they are not saved in the DB. I believe this might have been an issue in the master branch as well, but I couldn't properly test it due to an infinite loop I was having (which was fixed only recently).
* Finally I found that if the "Do not keep activities" is selected and I go into the detail page of a plugin and come back to the list, the app was crashing. This is because the `ViewModel` is also destroyed with the Activity. I've opened #7237 to handle this case separately where you can find more information about it.

Thanks a lot for the reviews!

nbradbury commented:
@oguzkocer This all looks good to me, but that "don't keep activities" issue has me concerned. It looks like we can add a one-line fix for this by using this in the activity's `onCreate()`:

```
if (savedInstanceState == null || mViewModel.getSite() == null) {
    mViewModel.setSite((SiteModel) getIntent().getSerializableExtra(WordPress.SITE));
}
```

oguzkocer commented:
> This all looks good to me, but that "don't keep activities" issue has me concerned. It looks like we can add a one-line fix for this by using this in the activity's onCreate():

@nbradbury That'll only fix the crash, but not the re-creation issue. Since we will be merging this to a WIP branch, I think it's OK to separate that PR which I'll work on today. Does that work for you?

nbradbury commented:
@oguzkocer That works for me, I'm just anxious to get the plugin browser merged into `develop` :)
